Resolve VID's to local store when sending message

chengr4 opened the pull request:
Fix part of #130 

> I need help guiding me to the right pathway

This is the fix for `tsp send`. For `tsp request`, an another PR is needed.

My question is: what method can we get 
- DID doc info
- validation info 
- and transport info ?

michielp1807 commented:
Thanks for your contribution!

After verifying a DID, the information related to this DID is stored in the wallet for later use. We shouldn't need to handle this information in the CLI application as this is done by the SDK.

There's seems to be some misunderstanding about the purpose of the `--ask` option. When a user uses `--ask`, the CLI should ask the user whether they want to verify the DID when a message from an unknown sender DID is received. This means the user can decide whether they want to add the new DID to their wallet. If `--ask` is not used, it should automatically verify when needed.

We only need to verify the sender, as we should already have the private details for the receiver DID, otherwise we can't decrypt the message.

Hopefully that makes sense, let us know if you have any more questions. Thanks!

chengr4 commented:
Hi @michielp1807,

Then I feel we need to implement one more method such as `has_verified_vid` of `AsyncSecureStore` in our SDK.

michielp1807 commented:
I agree, that would be a useful method to have.

I created a separate PR to add a `has_verified_vid` method, see #150. Feel free to rebase on that PR while wait for it to be approved and merged.

chengr4 commented:
My guess is that the user is not allowed to send a message to an unverified receiver.

Therefore, if a user chooses to use `--ask` and rejects to verify an unknown receiver DID, I assume the CLI would terminate the process => is that correct?

---

I have also been thinking about the appropriate level at which DID verification should happen. I saw the following code `in AsyncSecureStore`:

```rust
/// async fn main() {
///     // alice wallet
///     let mut db = AsyncSecureStore::new();
///     let alice_vid = OwnedVid::from_file("../examples/test/alice/piv.json").await.unwrap();
///     db.add_private_vid(alice_vid).unwrap();
///     db.verify_vid("did:web:raw.githubusercontent.com:openwallet-foundation-labs:tsp:main:examples:test:bob", None).await.unwrap();
///
///     // send a message
///     let result = db.send(
///         "did:web:raw.githubusercontent.com:openwallet-foundation-labs:tsp:main:examples:test:alice",
///         "did:web:raw.githubusercontent.com:openwallet-foundation-labs:tsp:main:examples:test:bob",
///         Some(b"extra non-confidential data"),
///         b"hello world",
///     ).await;
```

Because of this, I believe the verification behavior should be handled at the CLI level rather than in AsyncSecureStore.

Please correct me if I’m wrong.

michielp1807 commented:
Indeed, if a user uses `--ask` and does not want to verify the new unknown DID, then we can't send or open the message so we should terminate.

As for on what level DID verification should happen, I think it would be fine to do it at CLI level for now. In the future, we may rework it to do verification on a lower level as part of an automatic relationship establishment step (see #146).

chengr4 commented:
@michielp1807 I was thinking to create `examples/tests` to add an integration test for `tsp send`. However, I found some bash scripts like `cli-test-direct.sh`, but I didn't see any CI pipeline relying on them. So I assume those are just for local checks, and I should create a new integration test for automated checks?

chengr4 commented:
Hi,

This PR is ready for review. I'd appreciate it if you could point out anything that needs improvement. At least one major issue still needs to be fixed before we can merge this PR:

My understanding of stdin is that since both the message (`tokio::io::stdin()`) and the prompt (`io::stdin()`) share the same stdin, the yes/no response for the prompt will also possibly be received by the message receiver.

Reproduction:

```bash
# cli-test-direct.sh

# ... same
# change the following part
echo "---- verify the address of the receiver"
# tsp --wallet marlon verify --alias marc "$DID_MARC"

echo "---- verify the address of the sender"
tsp --wallet marc verify --alias marlon "$DID_MARLON"

echo "---- wait 2 seconds and then send a message to the receiver"
sleep 2 && echo "y\nOh hello Marc" | tsp --wallet marlon send -s marlon -r "$DID_MARC" &
# ... same
```

The receiver will get

```bash
y
Oh hello Marc
```

Additionally, I have few concerns wish to discuss with you

1. We currently have both a `test` and a `tests` folder under `examples`. I think we should consider cleaning this up.
2. I found that the `tsp print` subcommand might be overridden by the built-in `print` command on my Mac's terminal. I'm wondering if `print` is the best choice for a subcommand name. (But this is just an example tool maybe not important)

chengr4 commented:
I made some update, please check it out.

> For the scripts, I would  not using the `--ask` flag such that you don't get any prompts and this won't be a problem.

I found out that I missed `--ask` in the Bash command, but you still understood my question — thank you for that!

```bash
echo "---- wait 2 seconds and then send a message to the receiver"
sleep 2 && echo "y\nOh hello Marc" | tsp --wallet marlon send -s marlon -r "$DID_MARC" --ask &
```

Also, interestingly, if I tested by using `assert_cmd`, the receiver would automatically ignore `y`. 
Though I do not understand why, I accept it 🫠.

> Maybe we can rename the `test` folder to `data` since it contains example data? If we do move these files we will also need to update the DIDs accordingly. E.g. `did:web:raw.githubusercontent.com:openwallet-foundation-labs:tsp:main:examples:test:alice` will become `did:web:raw.githubusercontent.com:openwallet-foundation-labs:tsp:main:examples:data:alice`

I will create a new PR to fix this.

michielp1807 reviewed (changes requested):
Thanks for the update!

There still seem to be some formatting and Clippy lints that need to be fixed.

Additionally, some of the new tests seem to have failed to receive the appropriate messages correctly. I'm guessing this could be due to a recent change in the SDK where we first sent out a relationship request if there is no established relationship between the two endpoints (as is specified by the TSP specification), which we did not do before, so there might be an extra message.

chengr4 commented:
Hi,

I have fixed the test. Since our `tsp receive` command uses the `--one` flag, it makes sense that the message is about to request the relation.

Additionally, I followed the `Clippy` suggestion from commit `a9dddfc50`. However, I noticed that the official `assert_cmd` examples still use `&`, like this:

```rust
Command::new("ls")
        .args(&["-l", "-a"]) // use &
        .unwrap();
```

So I’m a bit unsure which one we should follow.

I also refactored some code in `Commands::Receive`. Please take a look when you get a chance.

chengr4 commented:
I found a `timeout` API that allows us to stop the receive command after a few seconds. However, the assertion should result in a failure [1]. 

Although it does eventually receive "Oh hello Marc", I am unsure if this is the correct behavior, especially since the receiver does not verify the sender (though I might be wrong).

Please take a look and let me know what you think. Thank you!

[1] https://docs.rs/assert_cmd/latest/assert_cmd/cmd/struct.Command.html#method.timeout

michielp1807 reviewed (approved):
Thanks a lot for your collaboration! I think we can merge this now
